File: ria/java/gof/src/main/java/com/gof/pattern/singleton/Settings4.java
package com.gof.pattern.singleton;

public class Settings4 {

    /**
     * double checked locking 사용
     **/
    private static volatile Settings4 instance;

    private Settings4() {
    }

    public static Settings4 getInstance() {
        if (instance == null) {
            synchronized (Settings4.class) {
                if (instance == null) {
                    instance = new Settings4();
                }
            }
        }
        return instance;
    }
}

File: zeri/gof/src/main/kotlin/com/gof/pattern/singleton/javakotlin/Settings4.kt
package com.gof.pattern.singleton.javakotlin

/**
 * [멀티 쓰레드 환경에서 안전하게 구현하는 방법]
 * - 이른 초기화 사용
 */
class Settings4 private constructor() {
    companion object {
        private val instance: Settings4 = Settings4()

        fun getInstance(): Settings4 = instance
    }
}
